cannon/serialize/binary_test.go

```diff
@@ -0,0 +1,93 @@
+package serialize
+
+import (
+	"encoding/binary"
+	"io"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestRoundTripBinary(t *testing.T) {
+	dir := t.TempDir()
+	file := filepath.Join(dir, "test.bin")
+	data := &serializableTestData{A: []byte{0xde, 0xad}, B: 3}
+	err := WriteSerializedBinary(file, data, 0644)
+	require.NoError(t, err)
+
+	hasGzip, err := hasGzipHeader(file)
+	require.NoError(t, err)
+	require.False(t, hasGzip)
+
+	result, err := LoadSerializedBinary[serializableTestData](file)
+	require.NoError(t, err)
+	require.EqualValues(t, data, result)
+}
+
+func TestRoundTripBinaryWithGzip(t *testing.T) {
+	dir := t.TempDir()
+	file := filepath.Join(dir, "test.bin.gz")
+	data := &serializableTestData{A: []byte{0xde, 0xad}, B: 3}
+	err := WriteSerializedBinary(file, data, 0644)
+	require.NoError(t, err)
+
+	hasGzip, err := hasGzipHeader(file)
+	require.NoError(t, err)
+	require.True(t, hasGzip)
+
+	result, err := LoadSerializedBinary[serializableTestData](file)
+	require.NoError(t, err)
+	require.EqualValues(t, data, result)
+}
+
+func hasGzipHeader(filename string) (bool, error) {
```

Comment:
Yeah interesting question. So far this is only used in the one package so it can just live here but it's potentially more useful (the JSON tests just check the compressed data can't be directly read as json).